### PR TITLE
DM-3517: Attempt to make RTLD_* constant discovery more robust

### PR DESCRIPTION
**Pull request by timj**

### CODE REVIEW

File: python/lsstimport.py [hunk 1]
```diff
@@ -37,59 +40,62 @@
...
-        RTLD_NOW = lsst64defs.RTLD_NOW         # usually 0x00002
-    DLFLAGS = RTLD_GLOBAL|RTLD_NOW
-except ImportError:
-    sys.stderr.write(
-        "Could not import lsst64defs; please ensure the base package has been built (not just setup).\n"
-    )
+# For portability we try a number of different options
+# Unfortunately on a Mac running 2.7 RTLD_NOW is not available anywhere
+options = ('os', 'DLFCN', 'ctypes', 'dl', 'lsst64defs')
+for mod in options:
```
Comment (ktlim) [inline python/lsstimport.py]: How frequently is this code executed?  If it's invoked on many/all imports, running through a long list of alternatives might not be advisable.  If it's only invoked once, then it's fine.

Comment (timj) [inline python/lsstimport.py]: My understanding is that this happens once on first import of the `lsstimport` module. Currently the `import dl` always fails and `lsst64defs` is always loaded. With this patch `os` will be loaded and won't be useful unless you are running python3 (but is a common module that is likely to be needed later) and then on Linux `DLFCN` will be loaded which will then work. On a Mac running Python 2.x `RTLD_NOW` will always fail to load until the last possible moment.

I could put in prior knowledge and do a version check for python3 and always load `lsst64defs` (which probably has a bad name at that point as it's not 64-bit specific anything) in every other case. We have to create `lsst64defs.py` regardless as we do not know at build time the python variant that we will be using at run time (if we take as a given that scons may not be running the same python).

This all means that we could drop all the self-determination stuff and always ask `lsst64defs.py` for the answer....

Comment (timj) [inline python/lsstimport.py]: I can confirm that when you `import lsstimport` it will only load once regardless of how many `__init__.py` files are read.

File: SConstruct [hunk 2]
```diff
@@ -1,19 +1,26 @@
 # -*- python -*-
 
+# Importing "lsst.sconsUtils" will cause Python
+# to load the __init__.py which will cause lsstimport to
+# be loaded which will hopefully not fail with a fatal error
+# but will initially print out a warning that base has not
+# been built yet.
```
Comment (jdswinbank) [inline SConstruct]: The commit message says that this "will load and fail first time". The above says it will "hopefully not fail". Reading those, I'm a bit unclear what's actually going to happen, how deterministic it is, and what the consequences will be. Could you maybe reword a bit for clarity?

File: python/lsstimport.py [hunk 3]
```diff
@@ -11,25 +12,27 @@
...
+from __future__ import print_function
 import sys
 import imp
 import functools
+import importlib
 import os.path
 
-#List of extensions to set global flags.  Mayb need to be extended
-#for systems other than *nix and OSX.
+# List of extensions to set global flags.  Mayb need to be extended
```
Comment (jdswinbank) [inline python/lsstimport.py]: Typo ("mayb") -- not in your text, but since you're editing it...

File: python/lsstimport.py (hunk 1 above)
Comment (jdswinbank) [inline python/lsstimport.py]: My understanding (which I think @timj is confirming above) is that we'll always either get the correct answer from `os` or `DLFCN`, or you'll fall through to `lsst64defs`. Is there any point in checking `ctypes` and `dl`?

Comment (timj) [inline python/lsstimport.py]: I left `dl` in there because that's what the code did previously. I agree that it's pointless given it's deprecated and we won't be using a python old enough and in 32-bit mode for it to ever work.

I'll remove `dl`. I left `ctypes` in because it does give us `RTLD_GLOBAL`. Whether that's useful depends on whether `lsst64defs` should ever be used. I'll address that in another comment.

File: SConstruct (hunk 2 above)
Comment (timj) [inline SConstruct]: What I'm trying to say is that `lsstimport` should be designed such that attempting load it before `lsst64defs` is created should not be fatal. A lot of this is irrelevant if we abandon `lsst64defs`.
